clang/include/clang/AST/ASTContext.h

```diff
@@ -2540,10 +2543,17 @@ class ASTContext : public RefCountedBase<ASTContext> {
 
   /// Determine whether the given types \p T1 and \p T2 are equivalent.
   bool hasSameType(QualType T1, QualType T2) const {
-    return getCanonicalType(T1) == getCanonicalType(T2);
+    if (getCanonicalType(T1) == getCanonicalType(T2)) {
+      return true;
```

Comment:
Should this return `true` if the qualifiers are different?

Reply:
It compares QualType so if qualifiers are different, the condition will fail and it will go to else. The else does check for duplicates only if qualifiers are same.